{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "from collections import Counter, defaultdict\n",
    "from copy import deepcopy\n",
    "from glob import glob\n",
    "from pathlib import Path\n",
    "from tempfile import NamedTemporaryFile\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import PDB, AlignIO\n",
    "from Bio.PDB.Chain import Chain\n",
    "from Bio.PDB.DSSP import DSSP\n",
    "from Bio.PDB.Model import Model\n",
    "from Bio.PDB.Structure import Structure\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = PDB.MMCIFParser()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_files = glob(\"../data/utp-data/structures/*/fold_*_model_0.cif\")\n",
    "structures = {\n",
    "    Path(f)\n",
    "    .stem.lstrip(\"fold_\")\n",
    "    .rstrip(\"_model_0\"): parser.get_structure(\n",
    "        Path(f).stem.lstrip(\"fold_\").rstrip(\"_model_0\"), f\n",
    "    )\n",
    "    for f in model_files\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_aligns = AlignIO.read(\"../data/ucyn-a_enriched/ucyn-a_enriched_cobalt_cleaned.fa\", \"fasta\")\n",
    "structure_ids = set(structures.keys())\n",
    "\n",
    "c_term_start = 880\n",
    "c_term_end = 1010\n",
    "\n",
    "aligns = {}\n",
    "for align in all_aligns:\n",
    "    full_id = align.id\n",
    "    _id = \" \".join(align.description.split(\" \")[1:]).replace(\"-\", \"_\").lower()\n",
    "    if _id not in structure_ids:\n",
    "        continue\n",
    "    align.id = _id\n",
    "    align.description = \"\"\n",
    "    aligns[_id] = align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_term_structs = {}\n",
    "for seqid, align in aligns.items():\n",
    "    # translate the aligned c_term_start and c_term_end to unaligned residue indicies\n",
    "    start_idx = len(align.seq[:c_term_start].replace(\"-\", \"\"))\n",
    "    struct = structures[seqid]\n",
    "    c_term_residues = list(struct.get_residues())[start_idx:]\n",
    "    new_structure = Structure(seqid + \"_c_term\")\n",
    "    new_chain = Chain(\"A\")\n",
    "    new_model = Model(0)\n",
    "\n",
    "    for residue in c_term_residues:\n",
    "        new_chain.add(residue)\n",
    "\n",
    "    new_model.add(new_chain)\n",
    "    new_structure.add(new_model)\n",
    "    c_term_structs[seqid] = new_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_xml_motifs(file_path):\n",
    "    meme_xml = ET.parse(file_path)\n",
    "    return meme_xml\n",
    "\n",
    "\n",
    "def extract_scanned_sites(meme_xml):\n",
    "    sites = meme_xml.findall(\".//scanned_sites\")\n",
    "    scanned_sites = [\n",
    "        (tag.attrib[\"sequence_id\"], site.attrib)\n",
    "        for tag in sites\n",
    "        for site in tag.findall(\"scanned_site\")\n",
    "    ]\n",
    "    return scanned_sites\n",
    "\n",
    "\n",
    "meme_xml = parse_xml_motifs(\"../data/utp-data/meme/meme_gb.xml\")\n",
    "scanned_sites = extract_scanned_sites(meme_xml)\n",
    "sequences = defaultdict(list)\n",
    "for seq, motif in scanned_sites:\n",
    "    sequences[seq].append(motif)\n",
    "sequences = {\n",
    "    seq: sorted(motifs, key=lambda motif: int(motif[\"position\"]))\n",
    "    for seq, motifs in sequences.items()\n",
    "}\n",
    "motif_seqs = {\n",
    "    tag.attrib[\"id\"]: tag.attrib[\"name\"] for tag in meme_xml.findall(\".//motif\")\n",
    "}\n",
    "\n",
    "seq_names = {\n",
    "    tag.attrib[\"id\"]: tag.attrib[\"name\"] for tag in meme_xml.findall(\".//sequence\")\n",
    "}\n",
    "seq_motifs = {\n",
    "    seq_name: sequences[seq_id]\n",
    "    for seq_id, seq_name in seq_names.items()\n",
    "    if seq_id in sequences\n",
    "}\n",
    "seq_motifs = {\n",
    "    seq_name: sorted(motifs, key=lambda motif: int(motif[\"position\"]))\n",
    "    for seq_name, motifs in seq_motifs.items()\n",
    "}\n",
    "seq_motifs_with_pos = {\n",
    "    seq_name: [(motif[\"motif_id\"], int(motif[\"position\"])) for motif in motifs]\n",
    "    for seq_name, motifs in seq_motifs.items()\n",
    "}\n",
    "seq_motifs = {\n",
    "    seq_name.replace(\"-\", \"_\").lower(): \"+\".join([motif[\"motif_id\"].split(\"_\")[1] for motif in motifs])\n",
    "    for seq_name, motifs in seq_motifs.items()\n",
    "}\n",
    "\n",
    "# motif_structures = defaultdict(list)\n",
    "# for seqid, motifs in seq_motifs_with_pos.items():\n",
    "#     full_id = align.id\n",
    "#     _id = seqid.replace(\"-\", \"_\").lower()\n",
    "#     if _id not in c_term_structs:\n",
    "#         continue\n",
    "#     struct = c_term_structs[_id]\n",
    "#     motifs_ids = \"+\".join([motif.split(\"_\")[1] for motif, _ in motifs])\n",
    "#     motif_pos = [pos for _, pos in motifs]\n",
    "#     motif_structures[motifs_ids].append((struct, motif_pos))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interesting_motifs = {\"2+1+6+3+4\", \"2+1+3+4\"}\n",
    "\n",
    "secondary_structures = {}\n",
    "\n",
    "for seqid, struct in c_term_structs.items():\n",
    "    if seq_motifs.get(seqid, \"\") not in interesting_motifs:\n",
    "        continue\n",
    "    \n",
    "    path = f\"../data/utp-data/structures/c_term/{seqid}_c_term.pdb\"\n",
    "    io = PDB.PDBIO()\n",
    "    io.set_structure(struct)\n",
    "    io.save(path)\n",
    "    \n",
    "    # add CRYST1 record to the pdb file\n",
    "    with open(path, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "    with open(path, \"w\") as f:\n",
    "        for line in lines:\n",
    "            if line.startswith(\"ATOM\"):\n",
    "                break\n",
    "            f.write(line)\n",
    "        f.write(\"CRYST1    1.000    1.000    1.000  90.00  90.00  90.00 P 1           1\\n\")\n",
    "        for line in lines:\n",
    "            f.write(line)\n",
    "        \n",
    "\n",
    "    model = struct[0]\n",
    "    dssp = DSSP(model, path)\n",
    "    sec_struct = [dssp[d][2] for d in dssp.keys()]\n",
    "    secondary_structures[seqid] = sec_struct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB import Superimposer\n",
    "\n",
    "rmsd_results = {}\n",
    "c_term_structs_list = list((k, v) for k, v in c_term_structs.items() if seq_motifs.get(k, \"\") in interesting_motifs)\n",
    "\n",
    "for idx1, (seqid1, structure1) in enumerate(c_term_structs_list):\n",
    "    if seq_motifs.get(seqid1, \"\") not in interesting_motifs:\n",
    "        continue\n",
    "    for seqid2, structure2 in c_term_structs_list[idx1 + 1:]:\n",
    "        if seq_motifs.get(seqid2, \"\") not in interesting_motifs:\n",
    "            continue\n",
    "        if seqid1 != seqid2:\n",
    "            atoms1 = list(structure1.get_atoms())\n",
    "            atoms2 = list(structure2.get_atoms())\n",
    "            if len(atoms1) > len(atoms2):\n",
    "                atoms1 = atoms1[:len(atoms2)]\n",
    "            elif len(atoms2) > len(atoms1):\n",
    "                atoms2 = atoms2[:len(atoms1)]\n",
    "            superimposer = Superimposer()\n",
    "            superimposer.set_atoms(\n",
    "                atoms1,\n",
    "                atoms2,\n",
    "            )\n",
    "            rmsd = superimposer.rms\n",
    "            rmsd_results[(seqid1, seqid2)] = rmsd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the RMSD results dictionary to a distance matrix\n",
    "dist_matrix = np.zeros((len(rmsd_results), ))\n",
    "for idx, rmsd in enumerate(rmsd_results.values()):\n",
    "    dist_matrix[idx] = rmsd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dist_matrix.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform hierarchical clustering\n",
    "Z = linkage(dist_matrix, method='average')\n",
    "\n",
    "# Plot the dendrogram\n",
    "plt.figure(figsize=(10, 6))\n",
    "dendrogram(Z, labels=[s for s, _ in c_term_structs_list], leaf_rotation=90)\n",
    "plt.xlabel('Sequence ID')\n",
    "plt.ylabel('Distance')\n",
    "plt.title('Hierarchical Clustering of RMSD Results')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=[s for s, _ in c_term_structs_list]\n",
    "labels[int(np.random.choice(Z[:, 0]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare structure of sequences with the same motif combianation and comapre with the generated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load aligned structures\n",
    "model_files = glob(\"../data/utp-data/structures/c_term/aligned/*.cif\")\n",
    "structures = {\n",
    "    Path(f)\n",
    "    .stem.lstrip(\"fold_\")\n",
    "    .rstrip(\"_model_0\"): parser.get_structure(\n",
    "        Path(f).stem.lstrip(\"fold_\").rstrip(\"_model_0\"), f\n",
    "    )\n",
    "    for f in model_files\n",
    "}\n",
    "del structures['1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "residue_positions = {\n",
    "    seqid: np.stack([residue.center_of_mass() for residue in structures[seqid][0][\"A\"]]) for seqid in structures\n",
    "}\n",
    "\n",
    "RADIUS = 3\n",
    "# for each residue in each structure, count the number of residues within a certain radius (of all structures)\n",
    "# and plot the results\n",
    "num_residues_within_radius = {}\n",
    "for seqid, residues in residue_positions.items():\n",
    "    num_residues = len(residues)\n",
    "    num_residues_within_radius[seqid] = np.zeros(num_residues)\n",
    "    for other_seqid, other_residues in residue_positions.items():\n",
    "        if other_seqid == seqid:\n",
    "            continue\n",
    "        for idx, residue in enumerate(residues):\n",
    "            num_residues_within_radius[seqid][idx] += np.sum(np.linalg.norm(other_residues - residue, axis=1) < RADIUS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "MIN_NUM_RESIDUES = 10\n",
    "\n",
    "cropped_structures = {}\n",
    "for seqid, residues_counts in num_residues_within_radius.items():\n",
    "    # select region where the number of residues_counts within the radius is greater than MIN_NUM_RESIDUES\n",
    "    roi_start, roi_end = None, None\n",
    "    for idx, r in enumerate(residues_counts):\n",
    "        if r > MIN_NUM_RESIDUES and roi_start is None:\n",
    "            roi_start = idx\n",
    "        elif r < MIN_NUM_RESIDUES and roi_start is not None:\n",
    "            roi_end = idx\n",
    "            break\n",
    "    if roi_start is None:\n",
    "        continue\n",
    "    if roi_end is None:\n",
    "        roi_end = len(residues_counts)\n",
    "    \n",
    "    struct = structures[seqid]\n",
    "    roi_residues = list(struct[0][\"A\"])[roi_start:roi_end]\n",
    "    new_structure = Structure(seqid + \"_roi\")\n",
    "    new_chain = Chain(\"A\")\n",
    "    new_model = Model(0)\n",
    "    \n",
    "    for residue in roi_residues:\n",
    "        new_chain.add(residue)\n",
    "    \n",
    "    new_model.add(new_chain)\n",
    "    new_structure.add(new_model)\n",
    "    cropped_structures[seqid] = new_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import colorsys\n",
    "\n",
    "def generate_distinct_colors(n):\n",
    "    colors = []\n",
    "    for i in range(n):\n",
    "        hue = i / n\n",
    "        saturation = 0.5\n",
    "        value = 0.8\n",
    "        rgb = colorsys.hsv_to_rgb(hue, saturation, value)\n",
    "        hex_code = '#%02x%02x%02x' % (int(rgb[0] * 255), int(rgb[1] * 255), int(rgb[2] * 255))\n",
    "        colors.append(hex_code)\n",
    "    return colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import py3Dmol\n",
    "from io import StringIO\n",
    "from Bio.PDB.PDBIO import PDBIO\n",
    "\n",
    "v = py3Dmol.view(width=800, height=800)\n",
    "\n",
    "# colors = generate_distinct_colors(len(cropped_structures))\n",
    "for idx, (seqid, struct) in enumerate(cropped_structures.items()):\n",
    "    io = StringIO()\n",
    "    pdb = PDBIO()\n",
    "    pdb.set_structure(struct)\n",
    "    pdb.save(io)\n",
    "    pdb_str = io.getvalue()\n",
    "\n",
    "    # v.setStyle({'cartoon': {'style':'cartoon'}})\n",
    "    # v.addModel(pdb_str, \"pdb\", {\"style\": {\"cartoon\": {\"color\": colors[idx]}}})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create consensus structure\n",
    "# 1. take the longest sequence\n",
    "# 2. for each residue, in each sequence, find the corresponding residue in the longest sequence, if it is within RADIUS\n",
    "# 3. calculate center of mass for each residue\n",
    "\n",
    "longest_seqid = max(cropped_structures, key=lambda seqid: len(cropped_structures[seqid][0][\"A\"]))\n",
    "longest_residues = list(cropped_structures[longest_seqid][0][\"A\"])\n",
    "\n",
    "consensus_structure = PDB.Structure.Structure(\"consensus\")\n",
    "consensus_model = PDB.Model.Model(0)\n",
    "consensus_chain = PDB.Chain.Chain(\"A\")\n",
    "consensus_model.add(consensus_chain)\n",
    "consensus_structure.add(consensus_model)\n",
    "consensus_stderrs = []\n",
    "for residue in longest_residues:\n",
    "    pos = residue.center_of_mass()\n",
    "    closest_residues = []\n",
    "    for seqid, struct in cropped_structures.items():\n",
    "        if seqid == longest_seqid:\n",
    "            continue\n",
    "        positions = np.stack([res.center_of_mass() for res in struct[0][\"A\"]])\n",
    "        distances = np.linalg.norm(positions - pos, axis=1)\n",
    "        if np.min(distances) < RADIUS:\n",
    "            closest_residues.append(list(struct[0][\"A\"].get_residues())[np.argmin(distances)])\n",
    "    consensus_pos = np.mean([res.center_of_mass() for res in closest_residues], axis=0)\n",
    "    stdev = np.std([res.center_of_mass() for res in closest_residues], axis=0)\n",
    "    consensus_stderrs.append(stdev)\n",
    "    majority_residue = Counter([r.get_resname() for r in closest_residues]).most_common(1)[0][0]\n",
    "    residue_template = next((r for r in closest_residues if r.get_resname() == majority_residue)).copy()\n",
    "    residue_template.detach_parent()\n",
    "    trans_vector = consensus_pos - residue_template.center_of_mass()\n",
    "    residue_template.transform(np.eye(3, 3), trans_vector)\n",
    "    residue_template.id = residue.id\n",
    "    residue_template.segid = residue.segid\n",
    "    consensus_chain.add(residue_template)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "io = StringIO()\n",
    "pdb = PDBIO()\n",
    "pdb.set_structure(consensus_structure)\n",
    "pdb.save(io)\n",
    "pdb_str = io.getvalue()\n",
    "\n",
    "v = py3Dmol.view(width=800, height=800)\n",
    "v.addModel(pdb_str, \"pdb\", {\"style\": {\"cartoon\": {\"color\": \"blue\"}}})\n",
    "# add spehere for each residue with radius proportional to the standard deviation\n",
    "# for idx, residue in enumerate(consensus_chain):\n",
    "#     pos = residue.center_of_mass()\n",
    "#     stderr = consensus_stderrs[idx]\n",
    "#     v.addSphere({\"center\": {\"x\": float(pos[0]), \"y\": float(pos[1]), \"z\": float(pos[2])}, \"radius\": float(np.linalg.norm(stderr)), \"color\": \"red\", \"opacity\": 0.8})\n",
    "    \n",
    "# save to file\n",
    "with open(\"../data/utp-data/structures/c_term/aligned/consensus.pdb\", \"w\") as f:\n",
    "    f.write(pdb_str)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
